compiler crashes (`@intFromEnum`?)

hi, i wanted to reimplement wayland in zig (mostly to learn wayland and zig, maybe use it in an app). i really liked the idea of subsurfaces, and i wanted to make them be able to move across the screen, so i started implementing a function that takes two lists of subsurfaces and writes move and resize requests only when needed. i started writing that in tree.zig, and now the compiler started segfaulting

### Zig Version

0.11.0 (flatpak)

### Steps to Reproduce and Observed Behavior

```
git clone https://codeberg.org/two/zdg
cd zdg
./makewp.sh
zig build-obj tree.zig
```

exists on `0.12.0-dev.2475+dcaf43674`
trace:
```
panic: InternPool.getCoerced of empty_enum_value not implemented from enum_type to int_type
Analyzing wl.zig: wl.zig:WBuf.write__anon_4431
      %856 = dbg_block_begin()
      %857 = dbg_stmt(2, 9)
      %858 = ref(%848) token_offset:88:19 to :88:23
      %859 = dbg_stmt(2, 23)
      %860 = field_ptr(%858, "content") node_offset:88:19 to :88:31
      %861 = dbg_stmt(2, 31)
      %862 = dbg_stmt(2, 38)
      %863 = field_call(.auto, %860, "writer", [
        {
          %864 = break_inline(%863, %851)
        },
      ]) node_offset:88:19 to :88:45
      %888 = ref(%863) token_offset:90:13 to :90:14
      %865 = dbg_var_val(%863, "w")
      %866 = save_err_ret_index(%863)
      %867 = dbg_stmt(3, 9)
      %868 = block_comptime({
        %869 = decl_ref("std") token_offset:89:21 to :89:24
        %870 = field_ptr(%869, "io") node_offset:89:21 to :89:27
        %871 = dbg_stmt(3, 42)
        %872 = field_call(.compile_time, %870, "CountingWriter", [
          {
            %873 = decl_val("NullWriter") token_offset:89:43 to :89:53
            %874 = break_inline(%872, %873)
          },
        ]) node_offset:89:21 to :89:54
        %875 = break(%868, %872)
      }) node_offset:89:21 to :89:54
      %876 = alloc_mut(%868) node_offset:89:9 to :89:101
      %877 = opt_eu_base_ptr_init(%876) node_offset:89:57 to :89:101
      %878 = struct_init_field_ptr(%877, "bytes_written") node_offset:89:77 to :89:78
      %879 = store_node(%878, @zero) node_offset:89:77 to :89:78
      %880 = struct_init_field_ptr(%877, "child_stream") node_offset:89:96 to :89:99
      %881 = typeof(%880) node_offset:89:96 to :89:99
      %882 = elem_type(%881) node_offset:89:96 to :89:99
      %883 = struct_init_empty_result(%882) node_offset:89:96 to :89:99
      %884 = store_node(%880, %883) node_offset:89:96 to :89:99
      %885 = validate_ptr_struct_init({
        %878 = struct_init_field_ptr(%877, "bytes_written") node_offset:89:77 to :89:78
        %880 = struct_init_field_ptr(%877, "child_stream") node_offset:89:96 to :89:99
      }) node_offset:89:57 to :89:101
      %886 = dbg_var_ptr(%876, "emptyw")
      %887 = dbg_stmt(4, 9)
      %889 = dbg_stmt(4, 14)
      %890 = dbg_stmt(4, 23)
      %891 = field_call(.auto, %888, "writeInt", [
        {
          %892 = break_inline(%891, @u32_type)
        },
        {
          %893 = break_inline(%891, %853)
        },
        {
          %894 = decl_val("native_endian") token_offset:90:33 to :90:46
          %895 = break_inline(%891, %894)
        },
      ]) node_offset:90:13 to :90:47
      %896 = try(%891, {
        %897 = err_union_code(%891) node_offset:90:9 to :90:47
        %898 = dbg_stmt(4, 9)
        %899 = ret_node(%897) node_offset:90:9 to :90:47
      }) node_offset:90:9 to :90:47
      %900 = ensure_result_used(%896) node_offset:90:9 to :90:47
      %901 = dbg_stmt(5, 9)
      %902 = decl_val("writeStruct") token_offset:91:13 to :91:24
      %903 = dbg_stmt(5, 24)
      %904 = call(.auto, %902, [
        {
          %905 = break_inline(%904, %851)
        },
        {
          %906 = dbg_stmt(5, 38)
          %907 = dbg_stmt(5, 45)
          %908 = field_call(.auto, %876, "writer", []) node_offset:91:32 to :91:47
          %909 = break_inline(%904, %908)
        },
        {
          %910 = break_inline(%904, @null_value)
        },
        {
          %911 = break_inline(%904, %854)
        },
      ]) node_offset:91:13 to :91:59
      %912 = try(%904, {
        %913 = err_union_code(%904) node_offset:91:9 to :91:59
        %914 = dbg_stmt(5, 9)
        %915 = ret_node(%913) node_offset:91:9 to :91:59
      }) node_offset:91:9 to :91:59
      %916 = ensure_result_used(%912) node_offset:91:9 to :91:59
      %917 = dbg_stmt(6, 9)
      %918 = load(%876) node_offset:92:23 to :92:29
      %919 = dbg_stmt(6, 29)
      %920 = field_val(%918, "bytes_written") node_offset:92:23 to :92:43
      %921 = int(8)
      %922 = dbg_stmt(6, 44)
      %923 = add(%920, %921) node_offset:92:23 to :92:47
      %924 = dbg_var_val(%923, "len64")
      %925 = dbg_stmt(7, 9)
      %936 = block({
        %926 = decl_ref("std") token_offset:93:21 to :93:24
        %927 = dbg_stmt(7, 24)
        %928 = field_ptr(%926, "math") node_offset:93:21 to :93:29
        %929 = dbg_stmt(7, 29)
        %930 = dbg_stmt(7, 34)
        %931 = field_call(.auto, %928, "cast", [
          {
            %932 = break_inline(%931, @u16_type)
          },
          {
            %933 = break_inline(%931, %923)
          },
        ]) node_offset:93:21 to :93:46
        %934 = is_non_null(%931) node_offset:93:21 to :93:75
        %935 = condbr(%934, {
          %937 = optional_payload_unsafe(%931) node_offset:93:21 to :93:75
          %938 = break(%936, %937)
        }, {
          %939 = dbg_stmt(7, 54)
          %940 = dbg_stmt(7, 54)
          %941 = ret_err_value("Overflow") token_offset:93:67 to :93:75
        }) node_offset:93:21 to :93:75
      }) node_offset:93:21 to :93:75
      %942 = dbg_var_val(%936, "len")
      %943 = save_err_ret_index(%936)
      %944 = dbg_stmt(8, 9)
    > %945 = int_from_enum(%854) node_offset:94:29 to :94:46
      %946 = as_node(@u16_type, %945) node_offset:94:29 to :94:46
      %947 = dbg_var_val(%946, "opcode")
      %948 = dbg_stmt(9, 9)
      %949 = dbg_stmt(9, 14)
      %950 = dbg_stmt(9, 23)
      %951 = field_call(.auto, %888, "writeInt", [
        {
          %952 = break_inline(%951, @u16_type)
        },
        {
          %953 = break_inline(%951, %946)
        },
        {
          %954 = decl_val("native_endian") token_offset:95:37 to :95:50
          %955 = break_inline(%951, %954)
        },
      ]) node_offset:95:13 to :95:51
      %956 = try(%951, {
        %957 = err_union_code(%951) node_offset:95:9 to :95:51
        %958 = dbg_stmt(9, 9)
        %959 = ret_node(%957) node_offset:95:9 to :95:51
      }) node_offset:95:9 to :95:51
      %960 = ensure_result_used(%956) node_offset:95:9 to :95:51
      %961 = dbg_stmt(10, 9)
      %962 = dbg_stmt(10, 14)
      %963 = dbg_stmt(10, 23)
      %964 = field_call(.auto, %888, "writeInt", [
        {
          %965 = break_inline(%964, @u16_type)
        },
        {
          %966 = break_inline(%964, %936)
        },
        {
          %967 = decl_val("native_endian") token_offset:96:34 to :96:47
          %968 = break_inline(%964, %967)
        },
      ]) node_offset:96:13 to :96:48
      %969 = try(%964, {
        %970 = err_union_code(%964) node_offset:96:9 to :96:48
        %971 = dbg_stmt(10, 9)
        %972 = ret_node(%970) node_offset:96:9 to :96:48
      }) node_offset:96:9 to :96:48
      %973 = ensure_result_used(%969) node_offset:96:9 to :96:48
      %974 = dbg_stmt(11, 9)
      %975 = decl_val("writeStruct") token_offset:97:13 to :97:24
      %976 = dbg_stmt(11, 24)
      %977 = call(.auto, %975, [
        {
          %978 = break_inline(%977, %851)
        },
        {
          %979 = break_inline(%977, %863)
        },
        {
          %980 = validate_ref_ty(%977) token_offset:97:35 to :97:36
          %981 = ref(%848) token_offset:97:36 to :97:40
          %982 = dbg_stmt(11, 40)
          %983 = field_ptr(%981, "fds") node_offset:97:36 to :97:44
          %984 = break_inline(%977, %983)
        },
        {
          %985 = break_inline(%977, %854)
        },
      ]) node_offset:97:13 to :97:50
      %986 = try(%977, {
        %987 = err_union_code(%977) node_offset:97:9 to :97:50
        %988 = dbg_stmt(11, 9)
        %989 = ret_node(%987) node_offset:97:9 to :97:50
      }) node_offset:97:9 to :97:50
      %990 = ensure_result_used(%986) node_offset:97:9 to :97:50
      %991 = dbg_block_end()
      %992 = restore_err_ret_index(%855.none)
      %993 = break(%855, @void_value)
    For full context, use the command
      zig ast-check -t wl.zig

  in wl.zig: wl.zig:WBuf.write__anon_4431
    > %855 = block({%856..%993}) node_offset:87:81 to :87:81
  in tree.zig: tree.zig:wl_update
    > %318 = try(%285, {%319..%321}) node_offset:34:23 to :34:171
  in tree.zig: tree.zig:wl_update
    > %278 = condbr(%277, {%280..%323}, {%324}) node_offset:34:9 to :34:171
  in tree.zig: tree.zig:wl_update
    > %279 = block({%275..%278}) node_offset:34:9 to :34:171
  in tree.zig: tree.zig:wl_update
    > %242 = block({%243..%436}) node_offset:29:38 to :29:38
  in tree.zig: tree.zig:wl_update
    > %236 = block({%241..%437}) node_offset:29:5 to :29:10
  in tree.zig: tree.zig:wl_update
    > %233 = condbr(%232, {%238..%438}, {%439}) node_offset:29:5 to :29:10
  in tree.zig: tree.zig:wl_update
    > %234 = block({%228..%233}) node_offset:29:5 to :29:10
  in tree.zig: tree.zig:wl_update
    > %227 = loop({%234, %237}) node_offset:29:5 to :29:10
  in tree.zig: tree.zig:wl_update
    > %216 = block({%217..%443}) node_offset:27:118 to :27:118
  in tree.zig: tree.zig:main
    > %484 = try(%474, {%485..%487}) node_offset:43:5 to :43:54
  in tree.zig: tree.zig:main
    > %449 = block({%450..%491}) node_offset:39:21 to :39:21
  in /home/dr/zig/build/stage4/lib/zig/std/start.zig: start.zig:callMain
    > %2851 = is_non_err(%2850) node_offset:585:28 to :585:45
  in /home/dr/zig/build/stage4/lib/zig/std/start.zig: start.zig:callMain
    > %2853 = block({%2847..%2852}) node_offset:585:28 to :585:45
  in /home/dr/zig/build/stage4/lib/zig/std/start.zig: start.zig:callMain
    > %2844 = block({%2845..%2991}) node_offset:584:24 to :584:24
  in /home/dr/zig/build/stage4/lib/zig/std/start.zig: start.zig:callMain
    > %2755 = switch_block(%2753,
        else => {%2994..%3000},
        %2756 => {%2757..%2769},
        %2770 => {%2771..%2789},
        by_val %2790 => {%2791..%2841},
        %2842 => {%2843..%2993}) node_offset:570:5 to :570:11
  in /home/dr/zig/build/stage4/lib/zig/std/start.zig: start.zig:callMain
    > %2739 = block({%2740..%3004}) node_offset:569:22 to :569:22
  in /home/dr/zig/build/stage4/lib/zig/std/start.zig: start.zig:initEventLoopAndCallMain
    > %2462 = builtin_call(%2460, %2461, @empty_struct) node_offset:519:12 to :519:48
  in /home/dr/zig/build/stage4/lib/zig/std/start.zig: start.zig:initEventLoopAndCallMain
    > %2307 = block({%2308..%2469}) node_offset:497:41 to :497:41
  in /home/dr/zig/build/stage4/lib/zig/std/start.zig: start.zig:callMainWithArgs
    > %2084 = call(.auto, %2082, []) node_offset:469:12 to :469:38
  in /home/dr/zig/build/stage4/lib/zig/std/start.zig: start.zig:callMainWithArgs
    > %2047 = block({%2048..%2091}) node_offset:462:72 to :462:72
  in /home/dr/zig/build/stage4/lib/zig/std/start.zig: start.zig:posixCallMainAndExit
    > %1873 = builtin_call(%1870, %1871, %1872) node_offset:425:17 to :425:79
  in /home/dr/zig/build/stage4/lib/zig/std/start.zig: start.zig:posixCallMainAndExit
    > %1869 = field_call(nodiscard .auto, %1866, "exit", [
        {%1870..%1874},
      ]) node_offset:425:5 to :425:80
  in /home/dr/zig/build/stage4/lib/zig/std/start.zig: start.zig:posixCallMainAndExit
    > %1478 = block({%1479..%1877}) node_offset:365:49 to :365:49

/home/dr/zig/lib/std/debug.zig:434:22: 0x705d14c in panicExtra__anon_133849 (zig)
    std.builtin.panic(msg, trace, ret_addr);
                     ^
/home/dr/zig/lib/std/debug.zig:409:15: 0x6b5adc6 in panic__anon_113539 (zig)
    panicExtra(null, null, format, args);
              ^
/home/dr/zig/src/InternPool.zig:7043:24: 0x661ea22 in getCoerced (zig)
                    inline .array_type, .vector_type => |seq_type| seq_type.child,
                       ^
/home/dr/zig/src/Module.zig:5611:62: 0x618c6ff in getCoerced (zig)
/// For `anyframe`, use the `InternPool.Index.anyframe` tag directly.
                                                             ^
/home/dr/zig/src/Sema.zig:8583:53: 0x655e44d in zirIntFromEnum (zig)
        },
                                                    ^
/home/dr/zig/src/Sema.zig:1054:69: 0x61771df in analyzeBodyInner (zig)
            .int_from_enum                => try sema.zirIntFromEnum(block, inst),
                                                                    ^
/home/dr/zig/src/Sema.zig:5918:34: 0x6acef87 in resolveBlockBody (zig)
    parent_block: *Block,
                                 ^
/home/dr/zig/src/Sema.zig:5901:33: 0x6611d2f in zirBlock (zig)
        .want_safety = parent_block.want_safety,
                                ^
/home/dr/zig/src/Sema.zig:1598:49: 0x6186d44 in analyzeBodyInner (zig)
                }
                                                ^
/home/dr/zig/src/Sema.zig:922:30: 0x64625f8 in analyzeBody (zig)
    _ = sema.analyzeBodyInner(block, body) catch |err| switch (err) {
                             ^
/home/dr/zig/src/Module.zig:4587:21: 0x6154725 in analyzeFnBody (zig)
            error.ComptimeReturn => unreachable,
                    ^
/home/dr/zig/src/Module.zig:3272:40: 0x5ed132e in ensureFuncBodyAnalyzed (zig)
                };
                                       ^
/home/dr/zig/src/Sema.zig:32247:31: 0x6b73aed in ensureFuncBodyAnalyzed (zig)
        .referencer = block.src_decl,
```

--- a/src/Sema.zig
+++ b/src/Sema.zig
@@ -8577,12 +8577,6 @@ fn zirIntFromEnum(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError
     const operand = try sema.resolveInst(inst_data.operand);
     const operand_ty = sema.typeOf(operand);
 
-    if (try sema.resolveValue(operand)) |resolved_op| {
-        if (resolved_op.isUndef(mod)) {
-            return sema.failWithUseOfUndef(block, operand_src);
-        }
-    }
-
     const enum_tag: Air.Inst.Ref = switch (operand_ty.zigTypeTag(mod)) {
         .Enum => operand,
         .Union => blk: {
@@ -8606,6 +8600,8 @@ fn zirIntFromEnum(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError
     };
     const enum_tag_ty = sema.typeOf(enum_tag);
 
+    // TODO: use correct solution
+    // https://github.com/ziglang/zig/issues/15909
     if (enum_tag_ty.enumFieldCount(mod) == 0 and !enum_tag_ty.isNonexhaustiveEnum(mod)) {
         return sema.fail(block, operand_src, "cannot use @intFromEnum on empty enum '{}'", .{
             enum_tag_ty.fmt(mod),
@@ -8613,6 +8609,12 @@ fn zirIntFromEnum(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError
     }
 
     const int_tag_ty = enum_tag_ty.intTagType(mod);
+
+    if (try sema.resolveValue(operand)) |resolved_op| {
+        if (resolved_op.isUndef(mod)) {
+            return mod.undefRef(int_tag_ty);
+        }
+    }
 
     if (try sema.typeHasOnePossibleValue(enum_tag_ty)) |opv| {
         return Air.internedToRef((try mod.getCoerced(opv, int_tag_ty)).toIntern());
@@ -20861,6 +20863,7 @@ fn zirTagName(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError!Air
     if (enum_ty.enumFieldCount(mod) == 0) {
         // TODO I don't think this is the correct way to handle this but
         // it prevents a crash.
+        // https://github.com/ziglang/zig/issues/15909
         return sema.fail(block, operand_src, "cannot get @tagName of empty enum '{}'", .{
             enum_ty.fmt(mod),
         });

--- a/src/Sema.zig
+++ b/src/Sema.zig
@@ -8599,6 +8599,7 @@ fn zirIntFromEnum(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError
         },
     };
     const enum_tag_ty = sema.typeOf(enum_tag);
+    const int_tag_ty = enum_tag_ty.intTagType(mod);
 
     // TODO: use correct solution
     // https://github.com/ziglang/zig/issues/15909
@@ -8608,19 +8609,15 @@ fn zirIntFromEnum(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError
         });
     }
 
-    const int_tag_ty = enum_tag_ty.intTagType(mod);
-
-    if (try sema.resolveValue(operand)) |resolved_op| {
-        if (resolved_op.isUndef(mod)) {
-            return mod.undefRef(int_tag_ty);
-        }
-    }
-
     if (try sema.typeHasOnePossibleValue(enum_tag_ty)) |opv| {
         return Air.internedToRef((try mod.getCoerced(opv, int_tag_ty)).toIntern());
     }
 
     if (try sema.resolveValue(enum_tag)) |enum_tag_val| {
+        if (enum_tag_val.isUndef(mod)) {
+            return mod.undefRef(int_tag_ty);
+        }
+
         const val = try enum_tag_val.intFromEnum(enum_tag_ty, mod);
         return Air.internedToRef(val.toIntern());
     }

--- a/src/Sema.zig
+++ b/src/Sema.zig
@@ -8577,6 +8577,12 @@ fn zirIntFromEnum(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError
     const operand = try sema.resolveInst(inst_data.operand);
     const operand_ty = sema.typeOf(operand);
 
+    if (try sema.resolveValue(operand)) |resolved_op| {
+        if (resolved_op.isUndef(mod)) {
+            return sema.failWithUseOfUndef(block, operand_src);
+        }
+    }
+
     const enum_tag: Air.Inst.Ref = switch (operand_ty.zigTypeTag(mod)) {
         .Enum => operand,
         .Union => blk: {
@@ -8589,6 +8595,7 @@ fn zirIntFromEnum(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError
                     .{src},
                 );
             };
+
             break :blk try sema.unionToTag(block, tag_ty, operand, operand_src);
         },
         else => {
@@ -8598,6 +8605,12 @@ fn zirIntFromEnum(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError
         },
     };
     const enum_tag_ty = sema.typeOf(enum_tag);
+
+    if (enum_tag_ty.enumFieldCount(mod) == 0) {
+        return sema.fail(block, operand_src, "cannot use @intFromEnum on empty enum '{}'", .{
+            enum_tag_ty.fmt(mod),
+        });
+    }
 
     const int_tag_ty = enum_tag_ty.intTagType(mod);
 

--- a/src/Sema.zig
+++ b/src/Sema.zig
@@ -8606,7 +8606,7 @@ fn zirIntFromEnum(sema: *Sema, block: *Block, inst: Zir.Inst.Index) CompileError
     };
     const enum_tag_ty = sema.typeOf(enum_tag);
 
-    if (enum_tag_ty.enumFieldCount(mod) == 0) {
+    if (enum_tag_ty.enumFieldCount(mod) == 0 and !enum_tag_ty.isNonexhaustiveEnum(mod)) {
         return sema.fail(block, operand_src, "cannot use @intFromEnum on empty enum '{}'", .{
             enum_tag_ty.fmt(mod),
         });

--- a/test/cases/compile_errors/int_from_enum_undefined.zig
+++ b/test/cases/compile_errors/int_from_enum_undefined.zig
@@ -1,0 +1,22 @@
+export fn a() void {
+    const E = enum {};
+    var e: E = undefined;
+    _ = &e;
+    _ = @intFromEnum(e);
+}
+
+comptime {
+    const E = union(enum(u16)) {
+        a: u32,
+        b: u32,
+    };
+    const e: E = undefined;
+    _ = @intFromEnum(e);
+}
+
+// error
+// backend=stage2
+// target=native
+//
+// :14:22: error: use of undefined value here causes undefined behavior
+// :5:22: error: cannot use @intFromEnum on empty enum 'tmp.a.E'

--- a/test/cases/compile_errors/int_from_enum_undefined.zig
+++ b/test/cases/compile_errors/int_from_enum_undefined.zig
@@ -5,18 +5,8 @@ export fn a() void {
     _ = @intFromEnum(e);
 }
 
-comptime {
-    const E = union(enum(u16)) {
-        a: u32,
-        b: u32,
-    };
-    const e: E = undefined;
-    _ = @intFromEnum(e);
-}
-
 // error
 // backend=stage2
 // target=native
 //
-// :14:22: error: use of undefined value here causes undefined behavior
 // :5:22: error: cannot use @intFromEnum on empty enum 'tmp.a.E'

